adding fiber generation codes

javijv4 opened the pull request:
<!-- Give your pull request (PR) a descriptive name -->

## Current situation
Resolves <https://github.com/SimVascular/svMultiPhysics/issues/479>

## Release Notes 
* Added fiber generation codes in the utilities folder.

## Code of Conduct & Contributing Guidelines 
- [x] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

javijv4 commented on utilities/fiber_generation/src/templates/solver_bayer.xml:
```diff
@@ -0,0 +1,266 @@
+<Save_results_to_VTK_format> 1 </Save_results_to_VTK_format>
+<Name_prefix_of_saved_VTK_files> result </Name_prefix_of_saved_VTK_files>
+<Increment_in_saving_VTK_files> 1 </Increment_in_saving_VTK_files>
+<Save_results_in_folder> /Users/jjv/Research/CEPCRTProject/ct_1010/heat/fibers </Save_results_in_folder>
```
These paths are overwritten by the Python file with the paths specified there. The modified template files are copied to the folder where the volume mesh is, and that .xml folder is run, so it doesn't really matter what paths are in the template. 

It doesn't make sense to have my paths, though, I agree, so I changed it to no path, just a general name.

javijv4 commented on utilities/fiber_generation/src/templates/solver_doste.xml:
```diff
@@ -0,0 +1,492 @@
+	</Add_face>
+	<Add_face name="av" >
+		<Face_file_path> /Users/jjv/Research/CEPCRTProject/ct_1010/mesh-surfaces/BASE.vtp </Face_file_path>
+	</Add_face>
```
Nope, but it doesn't matter for functionality, since they get overwritten based on the face name. I changed them to make it clear, though!

javijv4 commented on utilities/fiber_generation/example/ot/doste.png:
done! I added the sheet-normal direction as well.

aabrown100-git commented on utilities/fiber_generation/README.md:
```diff
@@ -0,0 +1,42 @@
+
+### Updates to the old code
+* All operations are vectorized now.
+* The SVmultiphysics solver now solves a Laplace equation.
```
Can you make all these "svMultiPhysics" for consistency?

aabrown100-git commented on utilities/fiber_generation/src/templates/solver_bayer.xml:
```diff
@@ -0,0 +1,266 @@
+	</Output>
+
+	<Output type="Alias" >
+		<Temperature> Phi_BiV_D</Temperature>
```
How is Phi_BiV_D used?

javijv4 replied:
Good catch! It's not doing anything. That problem was in the original file I got to generate the Bayer fibers, so I am guessing it is an auxiliary problem for something else, not the fibers. I deleted it.

aabrown100-git commented on utilities/fiber_generation/src/FibGen.py:
```diff
@@ -0,0 +1,1084 @@
+    gPhi_EPI, gPhi_LV, gPhi_RV, gPhi_AB = loadLaplaceSolnBayer(laplace_results_file)
+
+
+    # Write the fiber directions to a vtu files
```
Is this comment correct?

javijv4 commented on utilities/fiber_generation/README.md:
```diff
@@ -0,0 +1,35 @@
+### Updates to the old code
+* All operations are vectorized now.
+* The SVmultiphysics solver now solves a Laplace equation.
+* In Bayer: For the bislerp interpolation, instead of using the correction described in Bayer et al. (that returns a discontinuity), the basis are flipped to maintain a coherent fiber direction (see function `generate_fibers_BiV_Bayer_cells` in `FibGen.py`).
```
Hmm.. do you refer to explaining all three of these points? 

Explaining the vectorization and the Bayer modification with math would probably require rewriting the methods of the Bayer paper again, which might be a bit overkill for a README? Maybe I can add those details to the documentation? 

For the Bayer bislerp one, I've been trying to look for a math or code explanation that explains why the formula returns this discontinuity, but I haven't found it. If I rewrite the methods of the Bayer paper in the documentation, then I can indicate exactly where I performed the flip of the vectors. Would that work? 

Same for the Laplace, I think including math would be better suited for the documentation.

Let me know what you think

aabrown100-git replied:
Sounds good with the documentation! I don't think you need to explain the vectorization or the Laplace equation part. Just explain how this implementation is different from what's published in the Bayer paper

javijv4 commented on utilities/fiber_generation/README.md:
```diff
@@ -0,0 +1,42 @@
+* All operations are vectorized now.
+* The SVmultiphysics solver now solves a Laplace equation.
+* In Bayer: For the bislerp interpolation, instead of using the correction described in Bayer et al. (that returns a discontinuity), the basis are flipped to maintain a coherent fiber direction (see function `generate_fibers_BiV_Bayer_cells` in `FibGen.py`).
+* In Bayer: The beta angles were not being included correctly. The second rotation was being applied respect the first vector (circumferential) when it should be respect to the second vector (longitudinal) (see function `generate_fibers_BiV_Bayer_cells` in `FibGen.py`).
```
Same as before, should I add it to the documentation?

aabrown100-git replied:
Yes add to documentation

aabrown100-git commented on utilities/fiber_generation/README.md:
```diff
@@ -0,0 +1,35 @@
+# SV-fibergen
+Python + SVmultiphysics codes for fiber generation. Two methods are implemented:
+* Bayer et al. (2012). [link](https://doi.org/10.1007/s10439-012-0593-5)
```
Sure a validation document sounds good!

ktbolt commented on utilities/fiber_generation/main_bayer.py:
```diff
@@ -0,0 +1,64 @@
+Created on 2025/11/21 20:38:14
+
+@author: Javiera Jilberto Vallejos 
+'''
```
@javijv4 There is no need have date and author. This will also need a license header.

ktbolt commented:
@javijv4 I think it will be ok to just leave the code where it is and we can figure out where to put it in future, not sure how the new Slicer SimVascular code will be organized.

The important thing is to implement a general Python package that can be imported, does not assume a solver or certain face names. 

@aabrown100-git What do you think ?

aabrown100-git commented:
@ktbolt @javijv4 I think implementing it as a Python package is a good idea! From what I can tell, FibGen.py is basically a Python package already. Do we just need a few changes to make it more user friendly?

@javijv4 If we want to remove the template .xml files, you can write Python code to generate the desired xml file and include this in FibGen.py. A few prompts to the AI should get it done.

@ktbolt what do you mean the package should not assume a solver?

javijv4 commented:
Yeah, the strings are used to identify a particular heart surface. "epi" = epicardium, "mv" = mitral valve, "tv" = tricuspid valve. I will use more descriptive names and look into using an Enum class. Thanks!

javijv4 commented on utilities/fiber_generation/main_bayer.py:
```diff
@@ -0,0 +1,123 @@
+    parser.add_argument("--outdir", default=outdir, help="Output directory (default: %(default)s)")
+    args = parser.parse_args()
+
+    svfsi_exec = args.svfsi_exec
```
done!

aabrown100-git commented on utilities/fiber_generation/main_doste.py:
```diff
@@ -0,0 +1,142 @@
+
+        # B = beta angle (this have an opposite sign to the Doste paper, 
+        # but it's because the longitudinal direction is opposite)
+        'BENDORV': 20,
```
Should this be 0? That's what is published in the Doste paper, as well as Piersanti.

javijv4 commented:
@ktbolt done!

@aabrown100-git and @dseyler, I modified the basis definition to ensure consistent alpha and beta angle definitions across methods and chambers, and modified Doste to assign 2/3 of the septum to the LV.

javijv4 commented:
@ktbolt, @aabrown100-git  done!
* Changed package names to lowercase with underscores
* Added an __init__.py. The examples can be run from the folder without installation, but the packages can be installed using `pip install -e .`. 
* Fixed the comment values

aabrown100-git commented on utilities/fiber_generation/fiber_generation/fiber_generator.py:
```diff
@@ -0,0 +1,764 @@
+            if name not in mesh.point_data:
+                raise KeyError(f"Field '{name}' not found in mesh point_data")
+            
+            # Scale to [0, 1] before gradient computation
```
Why is this necessary? It's a bit confusing that this function may modify the field itself.

aabrown100-git replied:
Okay that makes sense! But for some Laplace fields that don't go from 0 to 1 (e.g. for Doste), we're not normalizing them to their appropriate range. And that might affect 1 and 2? What about explicitly normalizing each Laplace field to its appropriate range (whether that's [0,1] or [-2,1]) outside of this gradient function?

aabrown100-git reviewed (approved):
@javijv4 This all looks good to me now, thanks for all your good work on this!

ktbolt commented on utilities/fiber_generation/main_doste.py:
```diff
@@ -0,0 +1,85 @@
+Created on 2025/11/21 20:38:14
+
+@author: Javiera Jilberto Vallejos 
+'''
```
@aabrown100-git I think any code in the repository needs a license.

javijv4 replied:
@ktbolt Is this the header I should add?
```
# SPDX-FileCopyrightText: Copyright (c) Stanford University, The Regents of the University of California, and others.
# SPDX-License-Identifier: BSD-3-Clause
```

ktbolt replied:
@javijv4 That's it !

javijv4 commented on utilities/fiber_generation/src/FibGen.py:
```diff
@@ -0,0 +1,1084 @@
+        x: array-like of shape (N, 3)
+    Returns:
+        np.ndarray of shape (N, 3) with row-wise normalized vectors.
+    """
```
done!

aabrown100-git commented:
@javijv4 @ktbolt Anything else to improve with this PR, or are we good to merge?

ktbolt reviewed (approved):
OK to merge.
